### PR TITLE
Measurement transformation framework.

### PR DESCRIPTION
**Pull request by smonkewitz**
Measurement algorithm plugins may now specify a "transformation" which will
convert their raw outputs into a calibrated form, suitable for database
ingestion etc.

### CODE REVIEW

File: python/lsst/meas/base/base.py [hunk 1]
```diff
@@ -262,6 +263,23 @@ def fail(self, measRecord, error=None):
                    % self.__class__.__name__)
         raise NotImplementedError(message)
 
+    @staticmethod
+    def getTransformClass():
+        """!
+        Get the measurement transformation appropriate to this plugin.
+
+        This returns subclass of MeasurementTransform, which may be
```
Comment (smonkewitz) [inline python/lsst/meas/base/base.py]: s/subclass/a subclass/

File: python/lsst/meas/base/base.py [hunk 2]
```diff
@@ -262,6 +263,23 @@ def fail(self, measRecord, error=None):
...
+        instantiated with details of the algorithm configuration and then
+        called with information about calibration and WCS to convert from raw
+        measurement quantities to calibrated units. Calibrated data is then
+        provided in a separate output table.
+
+        By default, we assume that no transformation is appropriate, and
+        therefore return NullTransform.
+        """
```
Comment (smonkewitz) [inline python/lsst/meas/base/base.py]: I think that PassThroughTransform might be a less surprising default - or is the expectation that by default you don't want to present users with uncalibrated data, and so all such should be removed prior to database ingest?

File: python/lsst/meas/base/transforms.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+When a transformer is called, it is handed a `SourceCatalog` containing the
+measurements to be transformed, a `BaseCatalog` in which to store results, and
+information about the WCS and calibration of the data. It may be safely
+assumed that both are contiguous in memory, thus a ColumnView may be used for
+efficient processing.
+
+Transformations can be defined in Python or in C++. Python code should inherit
+from `MeasurementTransform`, following its interface. C++ code inherits from
+`BaseTransform`. `TransformWrapper` wraps the C++ implementation so that it
+provides a uniform interface with Python.
```
Comment (smonkewitz) [inline python/lsst/meas/base/transforms.py]: I really like your documentation (and not just here).

File: python/lsst/meas/base/transforms.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    This is intended as the default for measurements for which no other
+    transformation is specified.
+    """
+    def __call__(self, inputCatalog, outputCatalog, wcs, calib):
+        pass
+
+
+class PassThroughTransform(MeasurementTransform):
+    """
+    Copy all fields named after the measurement plugin from input to output, wihtout transformation.
```
Comment (smonkewitz) [inline python/lsst/meas/base/transforms.py]: s/wihtout/without/

File: python/lsst/meas/base/transforms.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    def __call__(self, inputCatalog, outputCatalog, wcs, calib):
+        pass
+
+
+class TransformWrapper(object):
+    """
+    Wrap a C++-style transformation.
+
+    C++ code requires a `Control`, rather than a `Config`, object. This class
+    converts the latter to the former when the transformer is instantiated.
```
Comment (smonkewitz) [inline python/lsst/meas/base/transforms.py]: Do you mean invoked or called, rather than instantiated?

File: python/lsst/meas/base/wrappers.py [hunk 6]
```diff
@@ -116,9 +118,15 @@ def wrapAlgorithm(Base, AlgClass, factory, executionOrder, name=None, Control=No
...
-                            getExecutionOrder=staticmethod(getExecutionOrder)))
+    if TransformClass:
+        PluginClass = type(AlgClass.__name__ + Base.__name__, (Base,),
+                           dict(AlgClass=AlgClass, ConfigClass=ConfigClass, factory=staticmethod(factory),
+                                getExecutionOrder=staticmethod(getExecutionOrder),
+                                getTransformClass=staticmethod(lambda: TransformClass)))
+    else:
+        PluginClass = type(AlgClass.__name__ + Base.__name__, (Base,),
+                           dict(AlgClass=AlgClass, ConfigClass=ConfigClass, factory=staticmethod(factory),
+                                getExecutionOrder=staticmethod(getExecutionOrder)))
```
Comment (smonkewitz) [inline python/lsst/meas/base/wrappers.py]: Can't you avoid stuttering the common initialization code by building the dict beforehand and adding a key for getTransformClass only if necessary?

File: tests/testTransform.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        # NB the choice of algorithm and executionOrder is arbitrary
+        singleFrame, forced = measBase.wrapSimpleAlgorithm(measBase.PsfFluxAlgorithm,
+                                                           Control=measBase.PsfFluxControl,
+                                                           executionOrder=0.0, doRegister=False)
+        self.assertEqual(singleFrame.getTransformClass(), measBase.BasePlugin.getTransformClass())
+        # Unless we override
+        singleFrame, forced = measBase.wrapSimpleAlgorithm(measBase.PsfFluxAlgorithm,
+                                                           Control=measBase.PsfFluxControl,
+                                                           executionOrder=0.0, doRegister=False,
+                                                           TransformClass=measBase.PassThroughTransform)
```
Comment (smonkewitz) [inline tests/testTransform.py]: Looks like there should be a check that singleFrame.getTransformClass() is PassThroughTransform here.

File: python/lsst/meas/base/base.py (hunk 2 above)
Comment (jdswinbank) [inline python/lsst/meas/base/base.py]: I have no strong opinion on this, but the current behaviour is what was requested by @TallJimbo; perhaps he'll pitch in here.

Jim -- if it's not obvious from the above, the question is: what should the calibration transform framework do when applied to a measurement plugin which does not have a specific transform specified? I think you previously indicated that it should not copy any of the measurements; the alternative Serge is suggesting is to copy all of them.

File: python/lsst/meas/base/transforms.py (hunk 5 above)
Comment (jdswinbank) [inline python/lsst/meas/base/transforms.py]: I actually _did_ mean instantiated here, but the way it's written is pretty confusing, so you're quite right to call me on it. I'll make it clearer.

File: python/lsst/meas/base/transforms.py (hunk 4 above)
Comment (jdswinbank) [inline python/lsst/meas/base/transforms.py]: Done.

File: python/lsst/meas/base/base.py (hunk 1 above)
Comment (jdswinbank) [inline python/lsst/meas/base/base.py]: Done.

File: tests/testTransform.py (hunk 7 above)
Comment (jdswinbank) [inline tests/testTransform.py]: Done.

File: python/lsst/meas/base/wrappers.py (hunk 6 above)
Comment (jdswinbank) [inline python/lsst/meas/base/wrappers.py]: Done.

File: python/lsst/meas/base/base.py (hunk 2 above)
Comment (TallJimbo) [inline python/lsst/meas/base/base.py]: I think Serge is probably right here.  There will be many columns (flags in particular, though we may want to handle them specially) that will not need to be transformed, but there will be very few that should be excluded from the calibrated catalog entirely.

Comment (jdswinbank) [inline python/lsst/meas/base/base.py]: Done.

**Pull request merged**